[alarmdecoder] Initial contribution

bobadair opened the pull request:
This PR is for a new OH2 version of the OH1 alarmdecoder binding. It has been developed jointly with @billfor.

The original OH1 version of this binding was contributed by @berndpfrommer. A significant amount of code has been retained from the original binding, although it may not be immediately obvious due to the extensive changes and restructuring required to fit the OH2 binding model and to meet current code requirements.

The binding is fully functional, but has been through only limited testing.

Closes #7081

cpmeister commented:
Make sure to remove '[WIP]' from the title if you want reviewers to review.

bobadair commented:
Removed WIP from the title. We're still looking for more testers, but I think this is far enough along to get the review process started.

bobadair commented:
Ok. I think that completes the list of high-priority changes, and eliminates all of the "TODO" comments in the code. Unless the testers find bugs that need to be fixed, I will now stop pushing commits until the initial review has been done.

One dumb question: Since I've now re-written the serial bridge handler to use the OHC serial transport rather than using gnu.io directly, can I remove the "<feature>openhab-transport-serial</feature>" line from the feature.xml file?

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/KeypadHandler.java:
```diff
@@ -0,0 +1,214 @@
+
+    private @NonNullByDefault({}) KeypadConfig config;
+    private boolean singleAddress;
+    private Pattern validCommandPattern = Pattern.compile(ADCommand.KEYPAD_COMMAND_REGEX);
```
```suggestion
    private static final Pattern validCommandPattern = Pattern.compile(ADCommand.KEYPAD_COMMAND_REGEX);
```

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/LRRHandler.java:
```diff
@@ -0,0 +1,108 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // All channels are read-only, so ignore all commands.
```
You still need to handle a RefreshType command if possible.

bobadair replied:
I thought about this, but the AD device mainly just passively listens on the panel's serial bus and passes along the messages it sees. There is no way to query it for state, as I would want to do when receiving a RefreshType command. So I thought it best to just ignore RefreshType along with all other commands.

cpmeister replied:
If there is no way to handle a RefreshType then I guess you can just ignore it.

bobadair replied:
Ok. Thanks. I'm open to suggestions, but I can't really thing of an appropriate REFRESH action that the AD would support.

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/SerialBridgeHandler.java:
```diff
@@ -0,0 +1,166 @@
+    private @NonNullByDefault({}) SerialPort serialPort = null;
+    private int serialPortSpeed = 115200;
+
+    @Activate
```
you can remove this annotation, it serves no purpose here

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/RFZoneHandler.java:
```diff
@@ -0,0 +1,129 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // Does not accept any commands
```
not even a refresh command?

bobadair replied:
Same as above.

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/SerialBridgeHandler.java:
```diff
@@ -0,0 +1,166 @@
+        if (writer != null) {
+            logger.trace("Closing writer");
+            try {
+                // writer.flush();
```
remove

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/ADCommand.java:
```diff
@@ -0,0 +1,137 @@
+
+    @Override
+    public String toString() {
+        return new String(command);
```
```suggestion
        return command;
```

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/ADMessage.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public String toString() {
+        return new String(message);
```
```suggestion
        return message;
```

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/EXPMessage.java:
```diff
@@ -0,0 +1,66 @@
+            channel = Integer.parseInt(parts.get(1));
+            data = Integer.parseInt(parts.get(2));
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("EXP message contains invalid number: " + e.getMessage());
```
```suggestion
            throw new IllegalArgumentException("EXP message contains invalid number: " + e.getMessage(), e);
```

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/IntCommandMap.java:
```diff
@@ -0,0 +1,65 @@
+            try {
+                commandMap.put(Integer.parseInt(kvPair[0]), kvPair[1]);
+            } catch (NumberFormatException e) {
+                throw new IllegalArgumentException("Unable to parse integer in mapping");
```
```suggestion
                throw new IllegalArgumentException("Unable to parse integer in mapping", e);
```

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/IntCommandMap.java:
```diff
@@ -0,0 +1,65 @@
+@NonNullByDefault
+public class IntCommandMap {
+    private final Map<Integer, String> commandMap;
+    private final Pattern validCommandPattern = Pattern.compile(ADCommand.KEYPAD_COMMAND_REGEX);
```
```suggestion
    private static final Pattern validCommandPattern = Pattern.compile(ADCommand.KEYPAD_COMMAND_REGEX);
```

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/KeypadMessage.java:
```diff
@@ -0,0 +1,147 @@
+            this.status = ((upper & 0x1F) << 13) | ((nbeeps & 0x3) << 10) | lower;
+
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("keypad msg contains invalid number: " + e.getMessage());
```
```suggestion
            throw new IllegalArgumentException("keypad msg contains invalid number: " + e.getMessage(), e);
```

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/LRRMessage.java:
```diff
@@ -0,0 +1,78 @@
+            }
+            partition = p;
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("LRR msg contains invalid number: " + e.getMessage());
```
```suggestion
            throw new IllegalArgumentException("LRR msg contains invalid number: " + e.getMessage(), e);
```

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/RFXMessage.java:
```diff
@@ -0,0 +1,64 @@
+            serial = Integer.parseInt(parts.get(0));
+            data = Integer.parseInt(parts.get(1), 16);
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("RFX message contains invalid number: " + e.getMessage());
```
```suggestion
            throw new IllegalArgumentException("RFX message contains invalid number: " + e.getMessage(), e);
```

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,326 @@
+
+		<config-description>
+			<parameter name="serialPort" type="text" required="true">
+				<label>Serial or USB Port</label>
+				<description>The name of the serial port used to connect to the Alarm Decoder device</description>
```
```suggestion
				<label>Serial or USB Port</label>
				<context>serial-port</context>
				<description>The name of the serial port used to connect to the Alarm Decoder device</description>
```

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,326 @@
+
+		<config-description>
+			<parameter name="hostname" type="text" required="true">
+				<label>Host Name</label>
+				<description>The hostname of the Alarm Decoder device</description>
```
```suggestion
				<label>Host Name</label>
				<context>network-address</context>
				<description>The hostname of the Alarm Decoder device</description>
```

bobadair replied:
This parameter can be a hostname or IP address. Does the network-address context apply to both?

cpmeister replied:
that context applies to both

bobadair replied:
Ok. Cool. Added it.

bobadair commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/ADThingHandler.java:
```diff
@@ -0,0 +1,82 @@
+
+        } else if (bridgeStatusInfo.getStatus() == ThingStatus.OFFLINE) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
+            // thingOfflineNotify();
```
Lost it. :-)

bobadair commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/IPBridgeHandler.java:
```diff
@@ -0,0 +1,177 @@
+    private @NonNullByDefault({}) IPBridgeConfig config;
+
+    /** hostname for the alarmdecoder process */
+    private @Nullable String tcpHostName = null;
+    /** port for the alarmdecoder process */
+    private int tcpPort;
```
Those were an artifact from the V1 code. Removed.

bobadair commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/IPBridgeHandler.java:
```diff
@@ -0,0 +1,177 @@
+
+        // Start connection check job
+        if (connectionSuccess) {
+            logger.debug("Scheduling connection check job with interval {} minutes.", config.reconnect);
+            lastReceivedTime = new Date();
+            connectionCheckJob = scheduler.scheduleWithFixedDelay(this::connectionCheck, config.reconnect,
+                    config.reconnect, TimeUnit.MINUTES);
```
Good point. Done.

bobadair commented:
Thanks for your comments @cpmeister, and for your very speedy initial review!
I just pushed a commit that I think addresses most of your comments. A few that I had questions about I commented on above.

cpmeister commented:
> can I remove the <feature>openhab-transport-serial</feature> line from the feature.xml file?

Nope, you need to keep that

bobadair commented:
Ok @cpmeister . I pushed another commit which I think addresses the rest of your comments, and also a minor doc update. Let me know if you think there are any other changes that should be made.

bobadair commented:
What is the next step for this binding? Does it need another review?

cpmeister commented:
I typically do multiple reviews for the code. Each review can be vary from superficial to thorough depending on how much sleep I've gotten and how much time I have. Reviewing ends when there is nothing left I think you should change.

But please be patient because there are a lot of PRs to review and I only have so much time in a day.

bobadair commented:
No problem! I was just curious.

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/KeypadMessage.java:
```diff
@@ -0,0 +1,147 @@
+     * @param other KeypadMessage to compare against
+     * @return true if messages are equal, false if other is null or messages are not equal
+     */
+    public boolean equals(@Nullable KeypadMessage other) {
```
You should override the `.equals(Object obj)` method instead of overloading.

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/KeypadMessage.java:
```diff
@@ -0,0 +1,147 @@
+     * Returns a string containing the keypad text
+     */
+    public String getText() {
+        return new String(alphaMessage);
```
```suggestion
        return alphaMessage;
```

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/LRRMessage.java:
```diff
@@ -0,0 +1,78 @@
+        eventData = parts.get(0);
+        cidMessage = parts.get(2);
+
+        if (parts.size() == 4) {
+            reportCode = parts.get(3);
+        } else {
+            reportCode = "";
+        }
```
```suggestion
        reportCode = parts.size() == 4 ? parts.get(3) : "";
```

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/ADBridgeHandler.java:
```diff
@@ -0,0 +1,427 @@
+
+    protected void startMsgReader() {
+        synchronized (msgReaderThreadLock) {
+            msgReaderThread = new Thread(this::readerThread, "AD Reader");
```
Please make sure to set the thread as daemon.

bobadair replied:
Done.

bobadair commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/SerialBridgeHandler.java:
```diff
@@ -0,0 +1,163 @@
+                serialPort.setSerialPortParams(serialPortSpeed, SerialPort.DATABITS_8, SerialPort.STOPBITS_1,
+                        SerialPort.PARITY_NONE);
+                serialPort.setFlowControlMode(SerialPort.FLOWCONTROL_RTSCTS_IN | SerialPort.FLOWCONTROL_RTSCTS_OUT);
+                // serialPort.disableReceiveFraming();
+                // serialPort.disableReceiveThreshold();
```
The original V1 code using gnu.io (RXTX) called these here, but org.eclipse.smarthome.io.transport.serial does not seem to provide them for some reason. I assume they default to off? I removed them and added a different comment as a reminder.

bobadair commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/SerialBridgeHandler.java:
```diff
@@ -0,0 +1,163 @@
+        try {
+            if (!serialDeviceName.isEmpty()) {
+                // Exit if no identifiers exist to work around possible bug
+                Stream<SerialPortIdentifier> serialPortIdentifiers = serialPortManager.getIdentifiers();
+                if (!serialPortIdentifiers.findAny().isPresent()) {
+                    logger.debug("No serial communication ports found. Cannot connect to [{}]", serialDeviceName);
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "No serial ports found");
+                    return;
+                }
```
Maybe. I moved it, but I'm hoping that we can get rid of it in 3.0 since the new serial library update. I added a comment to that effect. See openhab/openhab-core/#1426

bobadair commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/SerialBridgeHandler.java:
```diff
@@ -0,0 +1,163 @@
+                    return;
+                }
+
+                SerialPortIdentifier portIdentifier = serialPortManager.getIdentifier(serialDeviceName);
+                if (portIdentifier == null) {
+                    logger.debug("Serial Error: Port {} does not exist.", serialDeviceName);
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                            "Configured serial port does not exist");
+                    return;
+                }
```
I don't know. That depends on whether or not serial ports can by dynamically added/removed on a system. I thought it was safer to leave it in connect().

bobadair replied:
Ok. I moved it.

bobadair commented:
Ok. Pushed another commit to address your review comments. Thanks.

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/ADBridgeHandler.java:
```diff
@@ -0,0 +1,428 @@
+    /**
+     * Return the ZoneHandler for the given address and channel, or null if no handler exists and is initialized.
+     */
+    private @Nullable ZoneHandler findZoneHandler(int address, int channel) {
```
Right now you have this designed such that the bridge searches out handlers that should process a given message. This requires the bridge to have knowledge of all of the ADHandler subclasses and also a two step approach:
1. Bridge must identify which handlers could handle a message.
2. Bridge then passes on message to handlers found in step 1

I think it would be simpler if instead you just send the message to all of the handlers and let the handlers decide whether or not they want to handle it. This allows the bridge to be ignorant of the types of handlers as well as possibly allowing the bridge to be ignorant of the message type as well. Also instead of each handler having to implement 2 methods (one to judge if it can handle the message and another to actually handle it), each handler would only have to implement a single method which optionally handles the message.

bobadair replied:
Yeah, I'll admit I don't love the way this works at the moment. I'm not sure I really like the idea of sending every message to every handler either, though, although it would probably eliminate a lot of code.  The AD protocol can be very chatty, especially with keypad messages which it seems to send every few seconds. I'd rather not hit every zone/rfzone handler with every keypad message, since there can be a lot of both. Sending each message to every handler of the correct type may be ok, but that would still require maintaining or generating a list of the child handlers of each type. Originally I was planning on registering all of the child handlers in maps in the bridge so that look-ups would be very efficient, but that seemed complex and error-prone, especially given the address mask comparisons necessary for finding the appropriate keypad or lrr handler for a given message.
Given a choice, I would rather not mess with this for now, since it works well enough, and come up with a better scheme later.

bobadair replied:
Ok. I had some time today, so I rewrote this part of the code. The bridge now sends messages to all child handlers of the correct type, and they decide in their handleUpdate(ADMessage) methods whether or not they should act on them. I'll merge it in to this branch after a bit more testing.

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/ZoneHandler.java:
```diff
@@ -0,0 +1,122 @@
+        if (firstUpdateReceived.compareAndSet(false, true)) {
+            updateState(CHANNEL_CONTACT, OpenClosedType.CLOSED);
+        }
+
```
remove blank line

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/IntCommandMap.java:
```diff
@@ -0,0 +1,65 @@
+@NonNullByDefault
+public class IntCommandMap {
+    private final Map<Integer, String> commandMap;
+    private static final Pattern VALID_COMMAND_PATTERN = Pattern.compile(ADCommand.KEYPAD_COMMAND_REGEX);
```
please put static fields before non-static fields

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/LRRMessage.java:
```diff
@@ -0,0 +1,73 @@
+        try {
+            int p = 0;
+            try {
+                p = Integer.parseInt(parts.get(1));
+            } catch (NumberFormatException e) {
+                p = Integer.parseInt(parts.get(1), 16);
```
How would this is correct if you are given a hexadecimal number that doesn't have any letters in it?

bobadair replied:
Looking through the docs again, I think this field is unlikely to ever be a hex number. However, if it were, I believe it would have a 0x prefix. There is equivalent code in the keypad message parser, which comes from the V1 binding.

berndpfrommer replied:
I could not find precisely what the protocol specs say about that field. The specs don't even say it's an integer, but the example they are giving is:
"Event Data:
This is either the user number of the person performing the action or the zone that was bypassed."

http://www.alarmdecoder.com/wiki/index.php/Protocol#Format_5

I don't know, might be safest to treat it as a string? (Sorry, I don't receive LRR messages on my system, so I don't have any example messages).

bobadair replied:
Thanks @berndpfrommer. Yes, the AD protocol spec wasn't too much help. From looking at the SIA DC-05-1999.09 document (pp. 5-8) it seems clear that zone and user are both integers, but there is no clear indication as to what number system the AD would use to send them other than the example in there docs which appears to be decimal. Maybe @LuckyMallari would have some input on this?

Using a string could work as well if we can't come up with a definitive answer.  BTW - I've been planning to add another channel to the lrr thing that provides the whole LRR message payload as a string for better backward compatibility with the V1 binding, but I thought I would do that after this initial submission was merged.

LuckyMallari replied:
> Thanks @berndpfrommer. Yes, the AD protocol spec wasn't too much help. From looking at the SIA DC-05-1999.09 document (pp. 5-8) it seems clear that zone and user are both integers, but there is no clear indication as to what number system the AD would use to send them other than the example in there docs which appears to be decimal. Maybe @LuckyMallari would have some input on this?
> 
> Using a string could work as well if we can't come up with a definitive answer. BTW - I've been planning to add another channel to the lrr thing that provides the whole LRR message payload as a string for better backward compatibility with the V1 binding, but I thought I would do that after this initial submission was merged.

Sorry for being inactive on the thread.. busy with stuff ;) Had to dig from my old posts, but here's some info:

https://www.openhab.org/addons/bindings/alarmdecoder1/
```
LRR format from Alarm Decoder is:

!LRR:<data>,<partition>,<event>
```

and
```
Because <data> could be anything, it cannot be tied to a particular item as opposed to item bindings such as Contacts mentioned above. Instead, the entire message is sent to the binding.

!LRR:<entiremessage>
```

Now, if you look at the [history](http://www.alarmdecoder.com/wiki/index.php?title=LRR_Support&action=history) story of the page, it has not been updated since 2013, so my comments are still correct.

I am still using the AD binding v1, and still works flawlessly at the time of this writing. Here are some same LRR messages captured from my Vista20P:

```
!SER2SOCK Connected
!SER2SOCK SERIAL_CONNECTED
##1
!Sending.done
[1001000100000000----],008,[f70000ff1008001c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
#1
!Sending.done
!RFX:0746846,80
#
!Sending.done
1
!RFX:0746846,00
!Sending.done
[1001000100000000----],008,[f70000ff1008001c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
#2
!Sending.done
[0101050100000000----],008,[f70000ff1008050c0c020000000000],"ARMED ***AWAY***You                                                                              may exit now"
!LRR:000,1,ARM_AWAY
!Sending.done
!Sending.done
!Sending.done
!Sending.done
!Sending.done
[1001010100000000----],008,[f70000ff1008011c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
!RFX:0322328,84
[1001000100000000----],008,[f70000ff1008001c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
!LRR:003,1,OPEN
!Sending.done
!Sending.done
!LRR:000,1,ARM_STAY
[0011030100000000----],008,[f70000ff1008038c08020000000000],"ARMED ***STAY***You                                                                              may exit now"
!Sending.done
!Sending.done
!Sending.done
[0011000100000000----],008,[f70000ff1008008c08020000000000],"ARMED ***STAY***You                                                                              may exit now"
!Sending.done
!Sending.done
[1001010100000000----],008,[f70000ff1008011c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
[1001000100000000----],008,[f70000ff1008001c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
!LRR:003,1,OPEN
!Sending.done
!Sending.done
!Sending.done
!Sending.done
!Sending.done
[0011030100000000----],008,[f70000ff1008038c08020000000000],"ARMED ***STAY***You                                                                              may exit now"
!LRR:003,1,ARM_STAY
[0011000100000000----],008,[f70000ff1008008c08020000000000],"ARMED ***STAY***You                                                                              may exit now"
!Sending.done
!Sending.done
!Sending.done
!Sending.done
!Sending.done
[0011000100000000----],008,[f70000ff1008008c08020000000000],"ARMED ***STAY***You                                                                              may exit now"
!Sending.done
!Sending.done
!Sending.done
!Sending.done
!Sending.done
[1001010100000000----],008,[f70000ff1008011c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
!Sending.done
!Sending.done
!Sending.done
!Sending.done
[1001000100000000----],008,[f70000ff1008001c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
!LRR:002,1,OPEN
!Sending.done
!LRR:002,1,ARM_STAY
[0011030100000000----],008,[f70000ff1008038c08020000000000],"ARMED ***STAY***You                                                                              may exit now"
!Sending.done
!Sending.done
!Sending.done
!Sending.done
!Sending.done
!RFX:0702437,84
[0011000100000000----],008,[f70000ff1008008c08020000000000],"ARMED ***STAY***You may exit now"
!Sending.done
!Sending.done
!Sending..done
!RFX:0638432,80
!Sending.done
[1001010100000000----],008,[f70000ff1008011c08020000000000],"THE MALLARI'S     Ready to Arm  "
!RFX:0638432,00
[1001000100000000----],008,[f70000ff1008001c08020000000000],"THE MALLARI'S     Ready to Arm  "
!LRR:002,1,OPEN
```

Here's how to interpret the LRR messages:
* `!LRR:000,1,ARM_AWAY` - I simply sent # 2 (which is same as hitting # 2 on keypad. This means the system was Armed (away), but it is `000` because this is partition #. In my OH rules, I don't treat this as partition, on ARM/DISARM states, I treat `000` as "system" user, which means an open/arm was invoked without using a PIN (like doing a # 2) from keypad. 
* `!LRR:002,1,OPEN` - Disarmed by User 2
* `!LRR:003,1,ARM_STAY` - Armed stay by user 3. This is when you hit PIN then Stay (in Vista20P keypads, this is 3). For example, if user 3 pin code is 1234, then you get this by hitting 12343 on keypad.
* `!LRR:00,1,OPEN` - Disarmed by User 3


LRR Event Types can be retrieved [here](http://www.alarmdecoder.com/wiki/index.php/LRR_Support)

Note that LRR events will need to be enabled by user on the alarm itself. For example, for OPEN to show up in LRR, OPEN reports should be enabled (field *65)

You need to
* Filter (or switch/case) events such as OPEN/ARM_AWAY/ARM_STAY and treat the field as USER 
* Events such as LOWBAT/TROUBLE/etc and treat them as partition

I just dont have enough free time to test which events are user events and which are system/partition events

But to quickly answer your question, they are INTEGERS

bobadair replied:
Thanks @LuckyMallari ! That is great info. So we can assume that *data* and *partition* will always be decimal integers. I've removed the hex parsing code that @cpmeister originally commented on, and also added your LRR messages to the test data I'm using.

Do you know what determines whether the LRR messages will be of the format in your data or the "CID" format mentioned in the current protocol docs (e.g. "!LRR:012,1,CID_1441,ff")? The parsing code will work with either, so I only ask out of curiosity, and maybe for purposes of documentation. I see there is an AD firmware capability called "L2 - Long range radio messages (!LRR) with CID reporting format", so I've been assuming it is related to firmware version and/or configuration.

billfor replied:
I thought I read someplace that LRR is always in Ademco Contact ID format (CID) regardless of the alarm setting for the phoneline. 

The LRR seems to be another destination for reports. What is the benefit of LRR support in the binding if the same messages are sent to the keypad?

billfor commented:
Are all these changes necessary to accept the binding for now?  Bob is doing the world a favor by porting this v1 binding so that people with the Alarmdecoder have a way to keep using recent versions of openhab.  He doesn't even own an alarmdecoder.

Many of the issues that you point out, are already present in the v1 binding that was accepted. Bob is not responsible for those.  I've been running this binding for weeks, and the basic functionality works great -- I've stopped using the v1 binding entirely. 
Is it possible to commit these changes and then resolve the style/pattern issues in a future update?

bobadair commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/ADBridgeHandler.java:
```diff
@@ -0,0 +1,428 @@
+                            break;
+                    }
+                } catch (MessageParseException e) {
+                    logger.info("Error {} while parsing message {}", e.getMessage(), msg);
```
Well, at least for now, I would really like users to be warned of any parse errors so that they can report the problem. Otherwise there may be silent failures in some cases which we won't know about. Do you think it would make sense to leave it at INFO for now and bump it down to DEBUG later? Or I can bump it up to WARN for now if you would rather reserve INFO level.

bobadair commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/SerialBridgeHandler.java:
```diff
@@ -0,0 +1,162 @@
+                logger.debug("connected to serial port: {}", serialDeviceName);
+                panelReadyReceived = false;
+                startMsgReader();
+                updateStatus(ThingStatus.ONLINE);
```
It's actually the other way around. The ADBridgeHandler class contains all of the message sending, receiving, and parsing, and the subclasses (IPBridgeHandler & SerialBridgeHandler) contain mostly the IO setup, tear-down, retry and monitoring logic for their specific type of communication. The lines you indicate are shared between the two, and so could be abstracted out, but I wasn't sure it was really worth the effort given that they are wrapped in common exception handling, etc., along with the more subclass-specific connection setup code.

bobadair commented:
Ok @cpmeister. I've pushed a few commits over the past couple of days that I think address most of your comments. There is a bit of a side conversation going on about LRR messages, but I don't think it will have any implications for the code beyond the update I already made.

I don't know why the Travis CI build is failing and the Jenkins build is succeeding. I suspect it has nothing to do with the binding itself, but is probably related to the version change to 2.5.5-SNAPSHOT on 4/19. That always seems to cause problems.

bobadair commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/ZoneHandler.java:
```diff
@@ -0,0 +1,125 @@
+
+    private final Logger logger = LoggerFactory.getLogger(ZoneHandler.class);
+
+    private @NonNullByDefault({}) ZoneConfig config;
```
Ok. Easy enough to fix. It's a bit silly, though, isn't it? Now in every handler we're just creating a dummy config object in the constructor that is then immediately thrown away and replaced with the real one in the initialize() method. The other alternatives seem even more annoying, though.

cpmeister replied:
A little silly yes, but it is preferable to NPEs. The null checker may have it's annoyances but I've generally grown to like it since helps address possible concurrency issues as well as preventing NPEs. I just really wish it handled maps properly.

bobadair commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/KeypadMessage.java:
```diff
@@ -0,0 +1,154 @@
+
+        try {
+            int numeric = 0;
+            try {
+                numeric = Integer.parseInt(parts.get(1));
+            } catch (NumberFormatException e) {
+                numeric = Integer.parseInt(parts.get(1), 16);
+            }
```
I believe so. This comes from the original V1 binding code, so it has been running happily lo these many years. :-) I expect that in the rare instances this field contains a hex number, it will be prefixed with 0x so that the first parseInt will reliably throw an exception.
As a side note... who comes up with a protocol where an integer represented as text may be either decimal or hex? Sigh.

cpmeister replied:
> As a side note... who comes up with a protocol where an integer represented as text may be either decimal or hex? Sigh.

Probably not one person, but a group of people that don't check or correct each-other's work managed by a manager that doesn't care.

bobadair replied:
LOL. That's the sort of thing that slips through as a team approaches a release deadline.

bobadair commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/config/ZoneConfig.java:
```diff
@@ -0,0 +1,27 @@
+ */
+@NonNullByDefault
+public class ZoneConfig {
+    public @NonNullByDefault({}) Integer address;
```
Ok. Done for all of the config classes.

bobadair commented:
Thanks @cpmeister. I just pushed two new commits. The first addresses today's comments, and the second just bumps the version in pom.xml to 2.5.5.

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/ZoneHandler.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    @Override
+    protected void initDeviceState() {
```
Almost all of the of the handlers implement this method the same way, is it possible to have it refactored?

bobadair replied:
Yeah, I had actually meant to do that. Done.

bobadair commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/ADMsgType.java:
```diff
@@ -0,0 +1,66 @@
+ * @author Bernd Pfrommer - Initial contribution (OH1)
+ * @author Bob Adair - Re-factored and removed methods unused in OH2 binding
+ */
+public enum ADMsgType {
```
This is done, even though the comment isn't flagged as "Outdated".

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/IPBridgeHandler.java:
```diff
@@ -0,0 +1,165 @@
+        disconnect(); // make sure we are disconnected
+        writeException = false;
+        try {
+            socket = new Socket(config.hostname, config.tcpPort);
```
For instance, the warnings on the following two lines can be fixed like this:
```suggestion
            Socket socket = new Socket(config.hostname, config.tcpPort);
            this.socket = socket;
```

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/IPBridgeHandler.java:
```diff
@@ -0,0 +1,165 @@
+    protected synchronized void connectionCheck() {
+        logger.trace("Connection check job running");
+
+        if (msgReaderThread != null && !msgReaderThread.isAlive()) {
```
This is how you fix the warning here. It can be annoying, but it makes the code much more thread-safe. You get the general idea though, so I'll let you fix the rest.
```suggestion
        Thread msgReaderThread = this.msgReaderThread;
        if (msgReaderThread != null && !msgReaderThread.isAlive()) {
```

bobadair commented:
Ok. I just pushed a commit that eliminates all remaining compiler warnings. Strangely, not all of them were showing up in Eclipse.

bobadair commented:
Just saw PR #7449, so I ran spotless:apply and pushed another commit with the changes.

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/actions/BridgeActions.java:
```diff
@@ -0,0 +1,76 @@
+     * Reboot thing action
+     */
+    @RuleAction(label = "reboot", description = "Reboot the Alarm Decoder device")
+    public void reboot() {
+        if (bridge != null) {
```
This would solve the only warning left.
```suggestion
    public void reboot() {
        ADBridgeHandler bridge = this.bridge;
        if (bridge != null) {
```

bobadair replied:
How did I miss that one? That was one of that warnings that wasn't showing up in Eclipse.
Fixed.

bobadair commented:
Thanks!

J-N-K commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/IPBridgeHandler.java:
```diff
@@ -0,0 +1,174 @@
+        try {
+            Socket socket = new Socket(config.hostname, config.tcpPort);
+            this.socket = socket;
+            reader = new BufferedReader(new InputStreamReader(socket.getInputStream(), AD_CHARSET_NAME));
```
better set `StandardCharsets.UTF8` instead of the `UTF-8` string.

J-N-K commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/IPBridgeHandler.java:
```diff
@@ -0,0 +1,174 @@
+            connectionCheckJob = scheduler.scheduleWithFixedDelay(this::connectionCheck, config.reconnect,
+                    config.reconnect, TimeUnit.MINUTES);
+        } catch (UnknownHostException e) {
+            logger.debug("unknown hostname: {}", config.hostname);
```
no need to add debug logging, thing status changes are already logged

J-N-K commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/IPBridgeHandler.java:
```diff
@@ -0,0 +1,174 @@
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "unknown host");
+            disconnect();
+        } catch (IOException e) {
+            logger.debug("cannot open connection to {}:{} error: {}", config.hostname, config.tcpPort, e.getMessage());
```
see above

J-N-K commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/KeypadHandler.java:
```diff
@@ -0,0 +1,205 @@
+        updateState(CHANNEL_KP_ZONE, new DecimalType(kpm.getZone()));
+        updateState(CHANNEL_KP_TEXT, new StringType(kpm.alphaMessage));
+
+        updateState(CHANNEL_KP_READY, (kpm.getStatus(KeypadMessage.BIT_READY)) ? OnOffType.ON : OnOffType.OFF);
```
```suggestion
        updateState(CHANNEL_KP_READY, OnOffType.from(kpm.getStatus(KeypadMessage.BIT_READY)));
```

bobadair replied:
Nice. Didn't know about that method.

J-N-K commented on bundles/org.openhab.binding.alarmdecoder/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,330 @@
+
+		<config-description>
+			<parameter name="serialPort" type="text" required="true">
+				<label>Serial or USB Port</label>
```
```suggestion
				<label>Serial Or USB Port</label>
```

J-N-K commented on bundles/org.openhab.binding.alarmdecoder/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,330 @@
+			<parameter name="serialPort" type="text" required="true">
+				<label>Serial or USB Port</label>
+				<context>serial-port</context>
+				<description>The name of the serial port used to connect to the Alarm Decoder device</description>
```
```suggestion
				<description>The name of the serial port used to connect to the Alarm Decoder device.</description>
```

bobadair replied:
Well, this isn't actually a sentence, so I wouldn't normally put a period there. The parallel I would draw is sentence fragments in a bullet list. It's debatable, though. Do these description fields usually have periods?

J-N-K commented on bundles/org.openhab.binding.alarmdecoder/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,330 @@
+			</parameter>
+			<parameter name="discovery" type="boolean">
+				<label>Enable Discovery</label>
+				<description>Enable automatic discovery of zones and RF zones</description>
```
```suggestion
				<description>Enable automatic discovery of zones and RF zones.</description>
```

J-N-K commented on bundles/org.openhab.binding.alarmdecoder/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,330 @@
+
+		<config-description>
+			<parameter name="address" type="integer" required="true">
+				<label>Address</label>
+				<description>Zone address</description>
```
```suggestion
				<label>Zone Address</label>
```

If it's short enough, just use the label.

J-N-K commented on bundles/org.openhab.binding.alarmdecoder/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,330 @@
+				<description>Zone address</description>
+			</parameter>
+			<parameter name="channel" type="integer" required="true">
+				<label>Channel</label>
```
see above

J-N-K commented on bundles/org.openhab.binding.alarmdecoder/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,330 @@
+		<config-description>
+			<parameter name="addressMask" type="text" required="true">
+				<label>Address Mask</label>
+				<description>Keypad address mask</description>
```
```suggestion
				<description>Keypad address mask.</description>
```

J-N-K commented on bundles/org.openhab.binding.alarmdecoder/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,330 @@
+	<!-- Contact Channel Type -->
+	<channel-type id="contact-channel">
+		<item-type>Contact</item-type>
+		<label>Contact</label>
+		<description>Contact state</description>
```
```suggestion
		<label>Contact State</label>
```

Also below

J-N-K commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/config/RFZoneConfig.java:
```diff
@@ -0,0 +1,25 @@
+ */
+@NonNullByDefault
+public class RFZoneConfig {
+    public int serial = -1;
```
After reading the code below, this seems to be an invalid value. Leave it out if that is correct

bobadair replied:
Yes, I think that is a recent change. I initialize it to an invalid value (0 is potentially a valid value) rather than using a @Nullable Integer to determine if it was not set by the user. As a general rule, anything weird like this in the code was done to get rid of  null annotation warnings. I think here it isn't strictly necessary because that parameter should be marked as required, but it still seemed like a good idea to have the code enforce it.

cpmeister commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/ADBridgeHandler.java:
```diff
@@ -0,0 +1,372 @@
+            logger.trace("Notifying child handlers that panel is in ready state");
+
+            // Notify child zone handlers by calling notifyPanelReady() for each
+            for (Thing thing : getThing().getThings()) {
+                ADThingHandler handler = (ADThingHandler) thing.getHandler();
+                if (handler != null && (handler instanceof ZoneHandler || handler instanceof RFZoneHandler)) {
+                    handler.notifyPanelReady();
+                }
+            }
```
Technically `instanceof` will only return true for a non-null object, so you don't need to filter them out explicitly.

@J-N-K Also I think this suggested change is less readable than the original. 
In addition, it is less performant and includes the overhead of generating anonymous lambda classes in the bytecode.

Lambdas are fun and useful, but you shouldn't treat them as the only trick in your toolbox otherwise you will try to use them for everything.

J-N-K replied:
I think readability is probably a bit personal preference. But I‘m ok to leave it as-is.

bobadair commented on bundles/org.openhab.binding.alarmdecoder/doc/README_OH1.md:
```diff
@@ -0,0 +1,264 @@
+# Alarm Decoder Binding
```
This is the README file from the V1 binding. The new READMEfile references and links to it.

bobadair replied:
I think it is important for users to be able to easily reference it when upgrading from the V1 binding. I could link to it somewhere else, and so remove it from here, if I were sure that its location would not change after the 3.0 release. For example, maybe here: `https://github.com/openhab/openhab1-addons/blob/master/bundles/binding/org.openhab.binding.alarmdecoder/README.md`. However, I wasn't sure if this repository might be moved, renamed, or removed at some point.

kaikreuzer replied:
@bobadair I'd agree with @J-N-K that this should not be included here.
But the good news is that we have our website versioned and will keep the versions of older releases in place - so the link that you are looking for is https://www.openhab.org/v2.5/addons/bindings/alarmdecoder1/.

bobadair replied:
Ok. Cool. I have changed the README.md file to reference the V1 file at that URL and have removed the README_OH1 file from this branch.

bobadair commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/AlarmDecoderDiscoveryService.java:
```diff
@@ -0,0 +1,99 @@
+
+    private final Logger logger = LoggerFactory.getLogger(AlarmDecoderDiscoveryService.class);
+
+    private ADBridgeHandler bridgeHandler;
```
Yes. There are no SCA warnings.

bobadair commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/ADThingHandler.java:
```diff
@@ -0,0 +1,109 @@
+ * @author Bob Adair - Initial contribution
+ */
+@NonNullByDefault
+public abstract class ADThingHandler extends BaseThingHandler {
```
Yes. There are no SCA warnings.

bobadair commented on bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/config/SerialBridgeConfig.java:
```diff
@@ -0,0 +1,27 @@
+ */
+@NonNullByDefault
+public class SerialBridgeConfig {
+    public String serialPort = "";
```
LOL. I just changed this from being a @Nullable, It was previously annotated @NonNullByDefault({}). When I removed that, it caused too many null check warnings, so I went with this method instead. The handler will indicate a config error if it is unchanged.

J-N-K replied:
IMO `@Nullable` would be the correct way, but we can leave it as-is.

bobadair commented:
Thanks @J-N-K. I pushed two commits that I think address most of your comments, and I commented on some others. I'll have to take a look back through to see what I missed tomorrow. Maybe you'll have had a chance to look at the discovery code by then. There actually isn't very much to it.

bobadair commented:
> I have a general question regarding the discovery. If I understand it right there is some sort of message (RFX) received by the bridge (Where does this message come from? Is it triggered by something or sent by the device at some interval?) Which is then passed to the discovery service. It is not possible to trigger a scan for devices and the discovery service uses only the content of the message and the brigde-UID to create the discovered things, right?

Yes, exactly. There is, unfortunately, no way to query the AD device for zone information, so when REL or EXP messages (for zone things) or RFX messages (for rfzone things) are received by the bridge handler, it passes them along to the discovery code.  These all indicate state transitions for different types of zones. So, for example, when you first open a door or trigger a motion detector it would be discovered. It keeps hash sets of zones it has seen so it only "discovers" them once (per run, anyway). It is all just passive listening. There also does not appear to be any good way to discover the AD device itself.

> Then why do you need a seprerate service for each bridge? You could create one service, inject that into the handler factory and pass it to the bridge handlers on creation. Then - when the proper message is received you pass that message and the bridge-uid to the service. You could then remove the whole register/unregister logic from the handler factory and simplify the discovery service.

You're right. It could be done using a single discovery service, and that would simplify the handler factory. The downside would be that it would complicate the discovery code - not so much because it would have to deal with zone sets for multiple bridge UIDs as because it would have to deal with concurrency. Right now it doesn't, as it can only be called by the bridge's single reader thread. Since I think it will be rare for users to run more than one bridge at a time, it may be best to let sleeping dogs lie for now. WDYT?

bobadair commented on bundles/org.openhab.binding.alarmdecoder/README.md:
```diff
@@ -0,0 +1,267 @@
+## Discovery
+
+Background discovery is currently supported for **zone** and **rfzone** things.
+If the bridge *discovery* parameter is set to *true*, the first time a status message is seen from each zone or RF zone a corresponding thing will appear in the inbox.
```
These should all be fixed now in the commit I'm about to push.

bobadair commented:
Sorry, I've been doing some work on the Lutron binding, and switching my dev environment back and forth to modify and test this one has been slowing me down. Not to mention driving me nuts. Or more nuts than I already was? Anyway, I think the commit I just pushed should address a lot of the remaining comments.

bobadair commented:
Ok @J-N-K. I think I've finally addressed all of the comments that I didn't have questions about. Can you take a look? Thanks.

J-N-K commented on CODEOWNERS:
```diff
@@ -7,6 +7,7 @@
 # Add-on maintainers:
 /bundles/org.openhab.binding.airquality/ @kubawolanin
 /bundles/org.openhab.binding.airvisualnode/ @3cky
+/bundles/org.openhab.binding.alarmdecoder/ @bobadair @billfor @berndpfrommer
```
Did @berndpfrommer recently work on this binding? Since the CODEOWNERS is used to request reviewers, we should stick to those people that recenetly did major work on that code.

bobadair replied:
He is the author of the V1 binding, which this is derived from. I know he said he was short of time to work on an OH2 version, so I'm not sure if he even wants to see review notifications regarding it.  @berndpfrommer, do you want to be on this list?

bobadair commented:
> Please rebase on 2.5.x and run `mvn spotless:apply`. We don't want to introduce new codestyle errors.

I did a spotless:apply and committed the changes on April 22. Does it need to be run again?

J-N-K commented:
We did only change some XML formatters for comments. I guess that is not a problem in this PR. I wanted to approve and merge and therefore looked at the report from Jenkins:

```
05:37:54 [INFO] --- sat-plugin:0.9.0:report (sat-all) @ org.openhab.binding.alarmdecoder ---
05:37:54 [INFO] Individual report appended to summary report.
05:37:55 [ERROR] Code Analysis Tool has found: 
05:37:55  1 error(s)! 
05:37:55  1 warning(s) 
05:37:55  2 info(s)
05:37:55 [WARNING] org.openhab.binding.alarmdecoder.internal.handler.ADBridgeHandler.java:[367]
05:37:55 Classes/Interfaces should be annotated with @NonNullByDefault
05:37:55 [ERROR] .binding.alarmdecoder/pom.xml:[0]
05:37:55 The pom version is different from the parent pom version
```
You can ignore the pom-error, `2.5.5-SNAPSHOT` is the correct version. But can you fix the other warning (`@NonNullByDefault` for the bridge handler? Thanks.

bobadair commented:
> But can you fix the other warning (`@NonNullByDefault` for the bridge handler?

Thanks @J-N-K. I thought that code analyzer warning was invalid. It is on an exception class defined within ADBridgeHandler. When I annotate it with @NonNullByDefault I get the compiler warning "Nullness default is redundant with a default specified for the enclosing type ADBridgeHandler". What is the right way to get rid it? I thought I had read that SCA would no longer generate these warnings for inner classes.

J-N-K commented:
Mhm. Probably that is fixed in SAT 0.10.0 which is only available for the OH3 branch.

bobadair commented:
Thanks to everyone who tested this, commented on it, and reviewed the code!
